MAD-NG Features and Compatibility Modes

fsoubelet opened the pull request:
# MAD-NG Compatibility

This PR close #104 and closes #105. I will make a new issue for some exotic caveats (see caveat subsection below).

While users relying on default parameters will not be affected, I label this as a major version release as there is a substantial amount of behavioural changes.

## Main Changes

Full details relevant to the users can be found in the Changelog, details are given in sections below. Main items are:

- Handling of boolean and complex headers values (`MAD-NG` feature).
- Handling of bolean-type and complex-type columns (`MAD-NG` feature).
- Compatibility modes for dataframe validation (details below).

Some changes and fixes also show up:

- `TfsDataFrame` validation is now skipped on reading by default.
- `TfsDataFrame` validation is by default done in `MAD-X` compatibility mode before writing.
- Writing a dataframe with no headers (not empty headers) - e.g. a `pandas.DataFrame` - now works correctly.

## MAD-NG Features

Changes have been made to support boolean and complex values, both in headers and columns.
Special care is taken for the parsing of complex values given that the MAD-NG representation uses `i` for the complex part while Python uses `j`.

I am adding comments on the PR to help review.

### Caveat

Some of the more exotic features that can happen from `MAD-NG` are not supported, as discussed. These include for instance the overloaded ranges (i.e. `1..7`) and for these the user should use `pymadng` and do conversions in memory. I will create an issue on the repo to be tagged "wontfix" or similar.

Should a user encounter this, then an `UnknownTypeIdentifierError` is raised (see maintenance changes). I am open to directing them towards `pymadng` in the error message.

## Validation

The validation has been reworked and offers compatibility guarantees with `MAD-X` or `MAD-NG`. A valid mode should be provided to the `tfs.frame.validate` function, which accepts `MADX` / `MAD-X / `MADNG` / `MAD-NG` (case-insensitive).

Validation can still be skipped when reading or writing by providing `None`.

Details on the applied rules are given in the documentation so I won't duplicate them here. I'll point it out as PR comments.

## Added Tests

I have added an entirely new module for all tests related to validation.
I have also added a bunch of reader and writer tests to cove the new (`MAD-NG`) features.
I have added new **TFS** files in the inputs for each of these features, and I have added a compressed version (for each supported format) of a file containing them all to the relevant folder.

## Added Documentation

Docstrings have been updated everywhere relevant.

The documentation now holds a dedicated page detailing the TFS format (this closes #105).
There is also a dedicated page about validation and compatibility modes (this closes #104).

I heavily encourage downloading the built documentation from CI and having a look locally.

## Maintenance Changes

The most substantial maintenance change is the introduction of many new errors to be raised, which are each specific to a given problem one might encounter. They all inherit from the (previously unique) `TfsFormatError`, so users that were catching this one will not see their code broken.

I have made a slew of maintenance changes to the documentation: fixing warnings during generation, updating deprecated configuration parameters, fixed typos, tried to make a lot of things more explicitely explained, fixed broken links etc.

There are also a bunch of maintenance changes made to the tests. Mainly, I have moved quite a few common fixtures to the `conftest.py` file to avoid the duplications we had through files. I have tried to separate a bit more different testing parts.

I will add PR comments to the code to make it easy to determine what is new vs what is maintenance.

fsoubelet commented on tests/test_writer.py:
```diff
@@ -176,10 +172,46 @@ def test_no_validation_non_unique_columns(self, tmp_path):
+
+    # ----- Below are tests for files with MAD-NG features ----- #
+
+    def test_tfs_write_read_with_booleans(self, _tfs_dataframe_booleans, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_booleans, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_booleans, new, check_exact=False)  # float precision can be an issue
+
+    def test_tfs_write_read_with_complex(self, _tfs_dataframe_complex, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_complex, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_complex, new, check_exact=False)  # float precision can be an issue
+
+    def test_tfs_write_read_madng_like(self, _tfs_dataframe_madng, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_madng, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_madng, new, check_exact=False)  # float precision can be an issue
```
**FEATURE**

These are typical writer tests but use files for MAD-NG features.

fsoubelet commented on tests/test_writer.py:
```diff
@@ -299,17 +337,7 @@ def test_warning_on_non_unique_both(self, tmp_path, caplog):
         assert "Non-unique indices found" in caplog.text
 
 
-# ------ Fixtures ------ #
-
-
-@pytest.fixture
-def _tfs_dataframe() -> TfsDataFrame:
-    return TfsDataFrame(
-        index=range(3),
-        columns="a b c d e".split(),
-        data=np.random.rand(3, 5),
-        headers={"Title": "Tfs Title", "Value": 3.3663},
-    )
```
**MAINTENANCE**

Moved to conftest.

fsoubelet commented on tests/test_writer.py:
```diff
@@ -379,14 +407,5 @@ def _list_column_in_dataframe() -> TfsDataFrame:
     )
 
 
-@pytest.fixture
-def _pd_dataframe() -> pd.DataFrame:
-    return pd.DataFrame(
-        index=range(3),
-        columns="a b c d e".split(),
-        data=np.random.rand(3, 5),
-    )
-
-
```
**MAINTENANCE**

Moved to conftest.

fsoubelet commented on tfs/__init__.py:
```diff
@@ -23,7 +23,9 @@
 __all__ = [
     "concat",
     "read",
+    "read_tfs",
```
These were already exported due to lines 8-9 but not "nicely" exported.

fsoubelet commented on tfs/collection.py:
```diff
@@ -66,40 +66,40 @@ class TfsCollection(metaclass=_MetaTfsCollection):
-        ... class ExampleCollection(TfsCollection):
-        ...     beta = Tfs("beta_phase_{}.tfs")  # A TFS attribute
-        ...     other_value = 7  # A traditional attribute.
+        # All TFS attributes must be marked with the Tfs(...) class,
```
**MAINTENANCE**

All of this is to get rid of warnings when generating the documentation.

fsoubelet commented on tfs/errors.py:
```diff
@@ -5,6 +5,102 @@
 Errors that can be raised during the handling of **TFS** files.
 """
 
+from pathlib import Path
+
+# ----- Main Exception to Inherit From ----- #
```
**FEATURE**

All of these exceptions correspond to scenarios where _we were already raising an error_. The errors are just now more specific, and have tailored error messages.

fsoubelet commented on tfs/frame.py:
```diff
@@ -233,33 +236,48 @@ def validate(
             dataframe. Accepts `warn` and `raise` as values, case-insensitively, which dictates
             to respectively issue a warning or raise an error if non-unique elements are found.
     """
+    if compatibility.lower() not in VALIDATION_MODES:
+        errmsg = f"Invalid compatibility mode provided: '{compatibility}'."
+        raise ValueError(errmsg)
```
**Note:** in the `validate` function directly we expect `compatibility` to be one of the valid values. `None` is not accepted here, only in reader / writer functions (which will then skip calling `validate`).

fsoubelet commented on tfs/frame.py:
```diff
@@ -284,35 +301,69 @@ def _element_is_list(element):
-        raise TfsFormatError(errmsg)
+        raise SpaceinColumnNameError
+
+    # ----- Additional checks for MAD-X compatibility mode ----- #
+    if compatibility.lower() in ("madx", "mad-x"):
+        # Check that no boolean values are in the headers
+        if any(isinstance(header, bool) for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"Boolean values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Change their types in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain boolean values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that no complex values are in the headers
+        if any(isinstance(header, complex) for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"Complex values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Change their types in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain complex values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that the dataframe contains no boolean dtype columns
+        if any(pdtypes.is_bool_dtype(type_) for type_ in data_frame.dtypes):
+            LOGGER.debug(
+                f"Boolean dtype column found in dataframe {info_str}, which is incompatible with MAD-X."
+                "Change the column dtypes to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Dataframe can not contain boolean dtype columns in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that the dataframe contains no complex dtype columns
+        if any(pdtypes.is_complex_dtype(type_) for type_ in data_frame.dtypes):
+            LOGGER.debug(
+                f"Complex dtype column found in dataframe {info_str}, which is incompatible with MAD-X."
+                "Change the column dtypes or split it into real and imaginary values to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Dataframe can not contain complex dtype columns in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
```
**FEATURE**

All of these are new. They are what allows us to distinguish between MAD-NG and MAD-X modes.

Plainly, once the work of implementing boolean / complex in reader and writer are done, we just need to **not** accept them in `MAD-X` mode.

fsoubelet commented on tfs/reader.py:
```diff
@@ -33,43 +50,45 @@ def read_tfs(
     index: str | None = None,
     non_unique_behavior: str = "warn",
-    validate: bool = True,  # noqa: FBT001, FBT002
+    validate: str | None = None,
```
**NOTE**: We now skip validation by default when reading, as we assume the user trusts their TFS file is correct. This can always be turned on, but saves a lot of time on very big dataframes (such as sliced FCC lattices etc).

fsoubelet commented on tfs/reader.py:
```diff
@@ -125,25 +144,34 @@ def read_tfs(
     metadata: _TfsMetaData = _read_metadata(tfs_file_path)
```
As a reminder: this nicely takes care of reading just enough of the file to determine the headers, the column names and the columns types (plus some other things to optimize reading later on).

fsoubelet commented on tfs/reader.py:
```diff
@@ -125,25 +144,34 @@ def read_tfs(
-        raise TfsFormatError(errmsg)
+        raise AbsentColumnTypeError(tfs_file_path)
+
+    # The pandas engines do NOT support reading complex numbers, so we take note of which columns are
+    # complex-dtyped and will provide a specific converter function for pandas to use
+    dtypes_dict = dict(zip(metadata.column_names, metadata.column_types))
+    converters: dict[str, Callable] = {}
+
+    if np.complex128 in metadata.column_types:
+        LOGGER.debug("Complex columns detected, reading as strings and casting later")
+        for colname, dtype in zip(metadata.column_names, metadata.column_types):
+            if dtype is np.complex128:
+                converters[colname] = _parse_complex  # will a special converter
+                del dtypes_dict[colname]  # remove to avoid ParserWarning saying we provided both
```
**FEATURE**

Hopefully this is clear enough, but here goes:

- Keep track of which columns are complex-dtyped (from metadata reading and dtype identifiers - if the TFS file lies to us through the identifier we are fucked and will crash in the lines doing the reading below).
- Put these in a special dictionary (`converters`) and assign to that entry our custom-made function to load complex-dtyped data.
- Remove the entry from the first dictionnary (with all the "normal" columns).

After these lines we have:
- A `dtypes_dict` dict with all non-complex columns (name as key, dtype to use as value).
- A `converters` dict with all complex columns (name as key, function to use to load it as value).

These are then provided to the pandas reader which uses either its own functions for the loading or our implementation for the complex-columns (it does not have one).

fsoubelet commented on tfs/reader.py:
```diff
@@ -125,25 +144,34 @@ def read_tfs(
-        dtype=dict(
-            zip(metadata.column_names, metadata.column_types)
-        ),  # assign types at read-time to avoid re-assigning later
+        dtype=dtypes_dict,  # assign types at read-time to avoid conversions later
+        converters=converters,  # special handling for complex columns
```
And this is where we give the information to `pandas`.

fsoubelet commented on tfs/reader.py:
```diff
@@ -277,15 +311,40 @@ def _read_metadata(tfs_file_path: pathlib.Path | str) -> _TfsMetaData:
 
-def _parse_header(str_list: list[str]) -> tuple:
+def _parse_header(str_list: list[str]) -> tuple[str, bool | str | int | float]:
+    """
+    Expects a parsed header line content after the '@' header identifier,
+    as a .
+
+    Args:
+        str_list (list[str]): list of parsed elements from the header line
+            (we use parse in reader with 'shlex.split`).
+
+    Returns:
+        A tuple with the name of the header parameter for this line, as
+        well as its value cast to the proper type (as determined by the
+        type identifier).
+
+    Raises:
+        AbsentTypeIdentifierError: if no type identifier is found in the header line.
+        InvalidBooleanHeaderError: if the identifier type indicates a boolean
+            but the corresponding value is not an accepted boolean.
+    """
```
Maintenance

fsoubelet commented on tfs/reader.py:
```diff
@@ -277,15 +311,40 @@ def _read_metadata(tfs_file_path: pathlib.Path | str) -> _TfsMetaData:
-    name = " ".join(str_list[0:type_index])
-    value_string = " ".join(str_list[(type_index + 1) :])
-    return name, _id_to_type(str_list[type_index])(value_string.strip('"'))
+    if value_type is bool:  # special handling for boolean values
+        return name, _string_to_bool(value_string)
```
**FEATURE**

We have a special handling of booleans considering `Lua` / `MAD-NG` does not necessarily represent them as Python and we also accept `0`/`1`.

fsoubelet commented on tfs/reader.py:
```diff
@@ -277,15 +311,40 @@ def _read_metadata(tfs_file_path: pathlib.Path | str) -> _TfsMetaData:
-    return name, _id_to_type(str_list[type_index])(value_string.strip('"'))
+    if value_type is bool:  # special handling for boolean values
+        return name, _string_to_bool(value_string)
+    elif value_type is np.complex128:  # MAD-NG uses i for imaginary but Python needs j
+        value_string = value_string.replace("i", "j")
```
**FEATURE**

We have a special handling of complex values as `MAD-NG` writes them with `i` and not `j` for the complex part.

Note that if `j` was used in the file somehow we are still parsing correctly.

fsoubelet commented on tfs/tools.py:
```diff
@@ -52,18 +52,16 @@ def significant_digits(
-def remove_nan_from_files(
-    list_of_files: list[str | Path], replace: bool = False  # noqa: FBT001, FBT002
-) -> None:
+def remove_nan_from_files(list_of_files: list[str | Path], replace: bool = False) -> None:  # noqa: FBT001, FBT002
```
Maintenance

fsoubelet commented on tfs/writer.py:
```diff
@@ -30,11 +31,12 @@ def write_tfs(
     headerswidth: int = DEFAULT_COLUMN_WIDTH,
     non_unique_behavior: str = "warn",
-    validate: bool = True,  # noqa: FBT001, FBT002
+    validate: str = "madx",
```
**Note:** Validation is enabled in `MAD-X` compatility mode by default before writing (since it is the most restrictive).

fsoubelet commented on tfs/writer.py:
```diff
@@ -147,85 +167,137 @@ def _insert_index_column(data_frame: TfsDataFrame | pd.DataFrame, save_index: st
-    if type_str == "%s":
-        value = f'"{value}"'
-    return f"@ {name:<{width}} {type_str} {value:>{width}}"
+    value_str = TfsStringFormatter().format_field(value, _value_to_string_format_id(value))
```
**FEATURE**

This is where the work is done. The `TfsStringFormatter` takes care of the proper string representation, based on the dtype. See implementation below in the file.

fsoubelet commented on tfs/writer.py:
```diff
@@ -147,85 +167,137 @@ def _insert_index_column(data_frame: TfsDataFrame | pd.DataFrame, save_index: st
-    data_frame = _quote_string_columns(data_frame)
     data_frame = data_frame.astype(object)  # overrides pandas auto-conversion (lead to format bug)
-    return "\n".join(data_frame.apply(lambda series: format_strings.format(*series), axis=1))
+    string_formatter = TfsStringFormatter()
```
**FEATURE**

Same thing here, we let the formatter class handle everything for us.

fsoubelet commented on tfs/writer.py:
```diff
@@ -234,35 +306,115 @@ def _dtype_to_id_string(type_: type) -> str:
     if pdtypes.is_string_dtype(type_):
         return "%s"
-    errmsg = f"Provided type '{type_}' could not be identified as either a bool, int, float or string dtype"
+    if pdtypes.is_bool_dtype(type_):
+        return "%b"
```
We always write booleans in boolean form (true / false) and not numerical form (0 / 1).

fsoubelet commented on tfs/writer.py:
```diff
@@ -234,35 +306,115 @@ def _dtype_to_id_string(type_: type) -> str:
+
+    def format_field(self, value, format_spec):
+        if format_spec.endswith("b"):
+            return self._format_boolean(value, format_spec)
+
+        elif format_spec.endswith("c"):
+            return self._format_complex(value, format_spec)
+
+        if format_spec.endswith("s"):
+            return self._format_string(value, format_spec)
+
+        return super().format_field(value, format_spec)
+
+    def _format_boolean(self, value, format_spec: str):
+        """
+        Special case for booleans, as from their written version
+        will always be either 'true' or 'false' (lowercase) as
+        done by MAD-NG.
+        """
+        bool_str = str(bool(value)).lower()
+        return super().format_field(bool_str, f"{format_spec[:-1]}s")
+
+    def _format_complex(self, value, format_spec: str):
+        """
+        Special case for complex numbers, MAD-X and MAD-NG use 'i' for
+        the imaginary part and not 'j', so we have to replace that in.
+        """
+        return super().format_field(value, f"{format_spec[:-1]}g").replace("j", "i")
+
+    def _format_string(self, value, format_spec: str) -> str:
+        """
+        Special case as we need to ensure that strings are enclosed
+        in either " or ' quotes.
+        """
+        try:
+            if not value.startswith(('"', "'")):
+                value = f'"{value}"'
+        except AttributeError:
+            pass
+        return super().format_field(value, format_spec)
```
**FEATURE**

String representation is now handled here.

It expects the value and the format_spec, which we determine ahead of time in `_dtype_to_string_format_id` based on the dtype for the value.

Default handling from `string.Formatter` is done, and special cases are written for boolean and complex dtypes.

Note that this is not slower than the previous implementation for all previously accepted dtypes, actually a little faster :) Big thanks to @JoschD for the implementation.

fsoubelet reviewed (commented):
I am adding quite a few comments to help / guide the review, intended to be seen as you go through the changes.

Feel free to discard / resolve them as you go or when you're done.

fsoubelet commented:
@jgray-19 I have added you for review. Given there are quite a few complexities of the `tfs-pandas` internals at play I do not expect a full review but if you can have a look at the `MAD-NG` specifics (especially in the docs) and make sure there are no mistakes, it'd be lovely.

To get a build artefact of the documentation (very recommended):
- Got to Checks at the top of this page.
- Click on the Build and Upload Documentation` workflow (just to get to the summary).
- Click on the `site-build` to download.

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+
+    Validation is optional.
+    It is by default turned off at read-time, and turned on at write-time.
+    The default compability mode enforced before writing is ``MAD-X``, as this ensures the file would be accepted by both codes.
```
```suggestion
    The default compatibility mode enforced before writing is ``MAD-X``, as this ensures the file would be accepted by both codes.
```

fsoubelet replied:
Oh the irony, considering the file name was correct 🙄

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+    df = tfs.read("path/to/file.tfs")
+
+    # To validate with MAD-X compatibility
+    validate(df, compability="mad-x")  # or use "madx"
```
```suggestion
    validate(df, compatibility="mad-x")  # or use "madx"
```

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+    validate(df, compability="mad-x")  # or use "madx"
+
+    # To validate with MAD-NG compability
+    validate(df, compability="mad-ng")  # or use "madng"
```
```suggestion
    validate(df, compatibility="mad-ng")  # or use "madng"
```

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+    # To validate with MAD-NG compability
+    validate(df, compability="mad-ng")  # or use "madng"
+
+In case of compability issue, an exception is raised which will point to the specific incompatible element.
```
```suggestion
In case of compatibility issue, an exception is raised which will point to the specific incompatible element.
```

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+
+When validating a `TfsDataFrame`, the behavior in case one of these rules is violated depends on the value of the `non_unique_behavior` parameter.
+These rules are *always* checked against when validating a `TfsDataFrame`.
+Additional checks can be performed by setting the `compability` parameter, as described in the :ref:`MAD-NG <madng mode>` and :ref:`MAD-X <madx mode>` below.
```
```suggestion
Additional checks can be performed by setting the `compatibility` parameter, as described in the :ref:`MAD-NG <madng mode>` and :ref:`MAD-X <madx mode>` below.
```

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+MAD-X Compatibility
+-------------------
+
+The ``MAD-X`` compability mode is more restrictive, and enforces that none of the features listed in the :ref:`MAD-NG section <madng mode>` do not appear in the `TfsDataFrame`.
```
```suggestion
The ``MAD-X`` compatibility mode is more restrictive, and enforces that none of the features listed in the :ref:`MAD-NG section <madng mode>` do not appear in the `TfsDataFrame`.
```

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+
+.. admonition:: Default mode
+
+    The default compability mode enforced before writing is ``MAD-X``.
```
```suggestion
    The default compatibility mode enforced before writing is ``MAD-X``.
```

jgray-19 commented on doc/compatibility.rst:
```diff
@@ -0,0 +1,80 @@
+    # To validate with MAD-X compatibility
+    validate(df, compability="mad-x")  # or use "madx"
+
+    # To validate with MAD-NG compability
```
```suggestion
    # To validate with MAD-NG compatibility
```

jgray-19 commented on tfs/writer.py:
```diff
@@ -64,37 +66,42 @@ def write_tfs(
             to respectively issue a warning or raise an error if non-unique elements are found.
-        validate (bool): Whether to validate the dataframe before writing it to file. Defaults to ``True``.
+        validate (str): If an accepted value is given, validation will be performed before writing.
+            Defauts to `madx`, which will assert compatibility of the data with the ``MAD-X``code.
```
```suggestion
            Defauts to `madx`, which will assert compatibility of the data with the ``MAD-X`` code.
```

jgray-19 commented on doc/tfsformat.rst:
```diff
@@ -0,0 +1,142 @@
+- The table should not contain duplicate columns.
+- If spaces are present in strings, they should be enclosed in either single or double quotes.
+- The table data should not contain nested structures (lists, tuples, etc.).
+- The table data should not contain non-physical values (``NaN``, ``Inf``, etc.) as they would not be read back by either ``MAD-X`` or ``MAD-NG``.
```
Not relevant to the point you're getting across, but are we sure this is the case for MAD-NG? 
I think `nan` and `inf` might be accepted by MAD-NG - therefore it might be possible for MAD-NG to output a table with such values - untested. I can look into this, if you think this might cause issues.

Also, I assume `None` or the equivalent Lua `nil` is in the etc?

jgray-19 commented on doc/tfsformat.rst:
```diff
@@ -0,0 +1,142 @@
+================  ======================  =============== =========================
+
+It is also possible to include the length of the value into the type identifier, as in `%10s` for a string of length 10.
+`MAD-X` and `MAD-NG` inlcude the length in their output files, yet they are not mandatory to be able to read the files back into these programms. 
```
```suggestion
`MAD-X` and `MAD-NG` include the length in their output files, yet they are not mandatory to be able to read the files back into these programs. 
```
Or `programmes`?

jgray-19 commented on doc/tfsformat.rst:
```diff
@@ -0,0 +1,142 @@
+================  ======================  =============== =========================
+%s                                string        `%s name`  ``MAD-X`` and ``MAD-NG``
+%bpm_s                            string    `%bpm_s name`  ``MAD-X`` and ``MAD-NG``
+%d                        64-bit integer         `%d 174`  ``MAD-X`` and ``MAD-NG``
```
Looking at MAD-NG source code, I cannot see that %d is parsed in the headers - have you tested this? It doesn't matter in columns as it appears MAD-NG ignores the column type and parses item by item and just seems to convert to whatever it can, in the order number->complex->range->boolean->nil.

A little nuance: MAD-NG uses Lua 5.2, which does not have integers as such (https://www.lua.org/pil/2.3.html), so integers have a max of 52-bits (I believe this is the highest precision before its a float).

fsoubelet replied:
Oh, interesting. I have not used `MAD-NG` for work, do you think you can quickly try to load the `tests/inputs/madng.tfs` file and see if that goes fine?

fsoubelet replied:
Good catch. I just tried to have `MAD-NG` read the `tests/inputs/madng.tfs` file and it just cries with "invalid header field type: 'd'". I will have to work around that.

fsoubelet commented:
After consulting with Laurent:

- The expected length identifier for string headers was there for super old reasons and `MAD-NG` does not **actually** care for it. The check will be removed (and replaced by a check that an actual string dtype was provided as value) in the next `MAD-NG` release.
- The invalid `%d` identifier for integers was an oversight and will be handled in the next `MAD-NG` release.

In case of issues with users we can simply direct them to update their `MAD-NG`/`pymadng`.

I will still add some `pymadng` tests but we will have to wait for a new release of both `MAD-NG` and `pymadng` for these to work as expected, which pushes back the ETA of our release.

JoschD commented on tests/test_compression.py:
```diff
@@ -1,16 +1,17 @@
+# ----- Compression tests with 'classic' TFS files (no MAD-NG features) ----- #
 
 
 @pytest.mark.parametrize("extension", ["gz", "bz2", "zip", "xz", "zst", "tar", "tar.gz"])
```
make SUPPORTED_EXTENSIONS constant to be used in all `parametrize`

fsoubelet replied:
Ok, but as a constant in the file itself only. We basically support what `pandas` supports and they might change that so I don't want us to hardcode it into the package.

JoschD commented on tests/test_compression.py:
I think I would have created the compressed files from a single file on the fly via fixture, instead of adding them all to the repo

JoschD commented on tests/test_validation.py:
```diff
@@ -0,0 +1,135 @@
+            _ = read_tfs(_invalid_bool_in_header_tfs_file, validate=validation_mode)
+
+    @pytest.mark.parametrize("validation_mode", ["not ok", "ma-Dx", "nope", "madngg"])
+    def test_validation_raises_on_invalid_compatibility_mode(self, _tfs_dataframe, validation_mode):
```
with my suggestion of allowing falsy values as validation modes in reader/writer, maybe also add two test.raises for `_ = read_tfs` and `_ = write_tfs` into this test

fsoubelet replied:
Following our discussion and the change, only `None` is accepted to skip validation in reader and writer, and checks for validity are done by the validate function. Still want a test? Or did I misunderstand?

JoschD replied:
Yes, you can add this to the test that you already have that should raise the errors, only you call not only the validate function directly, but also the read and write function, which then should go into the validate function to raise the same error.

fsoubelet replied:
Oooh okay yes, will do.

fsoubelet commented on doc/tfsformat.rst:
```diff
@@ -0,0 +1,144 @@
+%b                               boolean        `%b true`           Only ``MAD-NG``
+%lz                      128-bit complex   `%lz 1.4+2.6i`           Only ``MAD-NG``
+================  ======================  =============== =========================
+
```
According to the MAD-X user guide if is for all (https://madx.web.cern.ch/madx/releases/last-rel/madxuguide.pdf). Have not seen it implemented anywhere like this for non-strings though, and I know MAD-NG also only writes it for strings.

fsoubelet commented on CHANGELOG.md:
```diff
@@ -1,5 +1,30 @@
+  - Handling of bolean-type and complex-type columns (`MAD-NG` feature).
+  - Compatibility modes for dataframe validation. The `tfs.frame.validate` function now expects this, and valid choices are `MADX`, `MAD-X`, `MADNG` and `MAD-NG` (case-insensitive, see API documentation).
+  - Many new exceptions have been created to be more specific when raising errors, which will now be more specific. They all inherit from the previously raised `TfsFormatError`, so user code that was catching it will still work.
+
```
I'm not sure what you mean? The new exceptions are each for a specific given issue and will point to this issue, but they all inherit from `TfsFormatError` so previous code going:
```python
try:
    some_stuff()
except TfsFormatError:
    other_stuff()
```
will work fine on the new version.

fsoubelet commented on tfs/reader.py:
```diff
@@ -125,25 +144,41 @@ def read_tfs(
+        for colname, dtype in zip(metadata.column_names, metadata.column_types):
+            if dtype is np.complex128:
+                converters[colname] = _parse_complex  # register it with our converter
+                del dtypes_dict[colname]  # remove to avoid ParserWarning saying we provided both
```
No, I added a commentto say why but maybe it's not clear enough.

If I don't delete from the first dict, then the column name is passed in two places to the pandas reader and the user gets a `ParserWarning` about it every time they load a file.

Should I make the comment more explicit?

fsoubelet replied:
Discussed: was for the comment.

JoschD commented on tfs/reader.py:
```diff
@@ -129,9 +130,9 @@ def read_tfs(
-
-        If one wants to, for instance, raise and error on non-unique indices or columns
-        when performing validation, one can do so as:
+        If one wants to raise an error on non-unique indices or columns
+        If one wants to, for instance, raise and error on non-unique indices or
```
duplicate now

fsoubelet replied:
thanks

fsoubelet commented on tfs/reader.py:
```diff
@@ -277,15 +318,40 @@ def _read_metadata(tfs_file_path: pathlib.Path | str) -> _TfsMetaData:
-def _parse_header(str_list: list[str]) -> tuple:
+def _parse_header(str_list: list[str]) -> tuple[str, bool | str | int | float]:
+    """
+    Expects a parsed header line content after the '@' header identifier,
+    as a .
```
Idk man, idk. Fixing now.

fsoubelet commented on tfs/reader.py:
```diff
@@ -277,15 +318,40 @@ def _read_metadata(tfs_file_path: pathlib.Path | str) -> _TfsMetaData:
+        return name, _string_to_bool(value_string)
+    elif value_type is np.complex128:  # MAD-NG uses i or I for imaginary but Python needs j
+        value_string = value_string.replace("i", "j").replace("I", "j")
+    return name, value_type(value_string)
```
thanks

fsoubelet commented on tfs/writer.py:
```diff
@@ -234,35 +306,115 @@ def _dtype_to_id_string(type_: type) -> str:
     raise TypeError(errmsg)
+
+
+class TfsStringFormatter(string.Formatter):
+    """Formatter class to be called for proper formatting of TFS strings."""
```
Well this is also called for the headers formatting, it's basically used everywhere for writing.

JoschD replied:
Yeah, but also there it formats values into strings. I'd say the nomenclature is header <--> data, not header <--> values. It sounds a bit like you are formatting strings, i.e. when the value itself is already a string.

fsoubelet commented:
This is shaping up quite well. Here are the few remaining things to do:

- [ ] Final conversation with Laurent
- [ ] Implement one or two tests writing a file and having `MAD-NG` read it (using `pymadng`), and making sure it goes fine
- [ ] Add to the documentation information about the first `MAD-NG` and `pymadng` releases we are compatible with (the one coming up this week brings `%s` and `%d` acceptance)

The last ones require waiting for the next `MAD-NG` and `pymadng` release so this PR is on hold until - most likely - later this week.

fsoubelet commented:
"Scope creep is not real, it can't hurt you"

I have added the support for `MAD-NG`'s `nil`s as we discussed. With this, apart from the tables and ranges that it includes (we recommend going through `pymadng` for these), we can confidently announce this release brings support for `MAD-NG`.

Caveats of the `nil` implementations (I have added these to the documentation too):
- When reading from file, `nil` values in the headers and string-dtyped columns are converted to `None` while those in the rest of the dataframe are cast to `NaN`.
- When writing to file, `None` values anywhere are written as `nil` and `NaN` values are written as `NaN` (remember that setting a `None` in a numeric `pandas.DataFrame` column automatically casts it as `NaN`, this is `pandas` behaviour).

I have disallowed `None` values in `MAD-X` compatibility modes. Relevant tests for reading, writing, validation regarding `nil`s are implemented.

- Only need left a `MAD-NG` & `pymadng` release so I can add tests for `MAD-NG` reading our files.

fsoubelet commented:
**The time has come.**

With the release of `MAD-NG` 1.0 and corresponding `pymadng` as well as the drop of Python 3.9 support, I believe I have finalized this PR for review. I understand this is a big review, I am available today for in-person discussion of the changes.

jgray-19 commented on doc/tfsformat.rst:
```diff
@@ -0,0 +1,142 @@
+
+.. code-block::
+
+    @ NAME                   %le             0.269975
```
Very pedantic, but would this ever be an example? Surely `NAME` would be a string?

fsoubelet replied:
Absolutely, good catch.

jgray-19 commented on doc/tfsformat.rst:
```diff
@@ -0,0 +1,142 @@
+- The table should not contain duplicate indices.
+- The table should not contain duplicate columns.
+- If spaces are present in strings, they should be enclosed in either single or double quotes.
+- The table data should not contain nested structures (lists, tuples, etc.).
```
What happens when a user has created a tfs dataframe through pymadng and then tries to use tfs.write(...). Will you error, warn or ignore?

In theory, you shouldn't do this. I.e. You can just use pymadng to read and write all these tfs dataframes, but what if someone tries...

fsoubelet replied:
I'll be honest, I have no idea exactly. Mostly likely is an error being raised.

I agree that in theory one should **not** do this. I don't particularly intend to cater to that use case. Eager for @JoschD's opinion

jgray-19 commented on tests/test_writer.py:
```diff
@@ -105,81 +107,159 @@ def test_write_int_float_columns(self, tmp_path):
+        # to ask to receive back from MAD-NG which would raise (and fail the test)
+        # if the loading did not go properly (our file is not accepted by MAD-NG).
+        madng = MAD(debug=True)
+        madng.protected_send(f"mtbl = mtable:read('{str(write_location.absolute())}')")
```
I think with my changes in 0.6.0, this is equivalent.
```suggestion
        madng.send(f"mtbl = mtable:read('{str(write_location.absolute())}')")
```

fsoubelet replied:
Indeed, fixed.

jgray-19 commented on tests/test_writer.py:
```diff
@@ -105,81 +107,159 @@ def test_write_int_float_columns(self, tmp_path):
+        # error loading it raises a `RuntimeError` and the test fails, otherwise it returns a
+        # np.int32(1) which we don't care about, and the test concludes successfully.
+        madng.send("py:send(1)")
+        madng.recv()
```
You don't check the contents after this??

fsoubelet replied:
Would love to. How would I go at doing this?

jgray-19 replied:
This might be useful: https://pymadng.readthedocs.io/en/latest/dataframes.html
My usage is here: https://github.com/pylhc/omc3/blob/6a9f229c3cea24c91c2666e0a15bc2200af5077b/tests/inputs/lhc_rdts/MAD_helpers.py#L286

fsoubelet replied:
Wonderful, thank you!

jgray-19 commented on tfs/frame.py:
```diff
@@ -116,17 +125,14 @@ def merge(
-            Any keyword argument is given to ``pandas.DataFrame.merge()``. The default values for all these
-            parameters are left as set in the ``pandas`` codebase. To see these, refer to the pandas
-            [DataFrame.merge documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html).
+            **kwargs: Any keyword argument is given to `pandas.DataFrame.merge`. The default
+                values for all these parameters are left as set in the ``pandas`` codebase.
```
This seems wordy...
ChatGPT suggests:
```suggestion
            **kwargs: Arguments for `pandas.DataFrame.merge`, with default values as
            defined in the ``pandas`` codebase.
```

jgray-19 commented on tfs/writer.py:
```diff
@@ -24,17 +25,18 @@
     """
-    Writes the provided ``DataFrame`` to disk at **tfs_file_path**, eventually with the `headers_dict` as
-    headers dictionary.
+    Writes the provided `DataFrame` to disk at **tfs_file_path**, eventually with
```
I don't understand... Why eventually?

fsoubelet replied:
The eventually refers to the `headers_dict` argument. If it's provided it will be used as headers, otherwise we use the headers of the frame if it has any, otherwise no headers. It's possible to write from a pandas dataframe and provide headers_dict, instead of constructing a `TfsDataFrame` first and writing. I'll rephrase a bit.

jgray-19 reviewed (commented):
Didn't look at any code functionality really, just the docs and docstrings.

jgray-19 commented on tests/test_writer.py:
```diff
@@ -105,81 +107,163 @@ def test_write_int_float_columns(self, tmp_path):
+        madng = MAD(debug=True)
+        madng.send(f"mtbl = mtable:read('{str(write_location.absolute())}')")
+
+        # If the loading went fine and we `.recv()` since we haven't sent anything it would
+        # just hang. So we first send some dummy and then call `.recv()`. Now if there was an
+        # error loading it raises a `RuntimeError` and the test fails, otherwise it returns a
+        # np.int32(1) which we don't care about, and the test concludes successfully.
+        madng.send("py:send(1)")
+        madng.recv()
```
I believe there is need for this now, since you communicate on 259...?

fsoubelet replied:
This is correct, patched.

JoschD reviewed (approved):
I think I read everything now. It's not easy as github does not tell me what has changed since my last view in all of these changes per file. But LGTM

JoschD commented on tfs/frame.py:
```diff
@@ -338,39 +341,39 @@ def _element_is_list(element):
+            errmsg = "TFS-Headers can not contain 'None' values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # MAD-X will not accept back in a TFS file with no 'TYPE' entry in the headers (as string)
+        if _has_headers(data_frame) and "TYPE" not in data_frame.headers:
+            LOGGER.warning("MAD-X expects a 'TYPE' header in the TFS file, which is missing. Adding it.")
+            data_frame.headers["TYPE"] = "Added by tfs-pandas for MAD-X compatibility"
```
This is not good. Not a valid dataframe if there is no header.
I would do something more like

```
headers = getattr(data_frame, {})
```

at the very beginning, then check everything on this object (no need for the `_has_headers()` function). Add TYPE to it here in the last check and then set `data_frame.headers = headers` at the end

fsoubelet replied:
> This is not good. Not a valid dataframe if there is no header.

We've been accepting it so far..

Also the validation function does not return anything. So if there are no headers, then we're adding "TYPE" to some dict which goes nowhere and we haven't helped the user.

fsoubelet commented:
Rounding error solved, failing right-alignment on strings in headers solved.

I think this can be the final review @JoschD @jgray-19 (famous last words, I know)

JoschD commented on tfs/writer.py:
```diff
@@ -64,56 +69,73 @@ def write_tfs(
         .. code-block:: python
 
-            >>> tfs.write("filename.tfs", dataframe, non_unique_behavior="raise")
+            tfs.write("filename.tfs", dataframe, non_unique_behavior="raise")
```
outdated, as one needs to activate validation first

fsoubelet replied:
Changed to how to DO validation if the user wants to.

JoschD commented:
Also: add a test for writing that takes a `pd.DataFrame` as well as one with `tfs.TfsDataFrame`  (empty header) and `headers_dict` and does validation.
(unless you already did and I didn't see it)

fsoubelet commented on tfs/writer.py:
```diff
@@ -64,56 +69,73 @@ def write_tfs(
-        One can choose to skip dataframe validation **at one's own risk** before writing
-        it to file. This is done as:
+        One can choose to skip dataframe validation at one's own risk before writing
+        it to file. This can be done as:
 
         .. code-block:: python
 
-            >>> tfs.write("filename.tfs", dataframe, validate=False)
+            tfs.write("filename.tfs", dataframe, validate=False)
```
Moved below the (new) example on how to DO the validation.

fsoubelet commented on tfs/writer.py:
```diff
@@ -64,56 +69,73 @@ def write_tfs(
+            Defauts to `madx`, which will assert compatibility of the data with the ``MAD-X`` code.
+            If `madng` is given then compatibility with the ``MAD-NG`` code is checked, which has more
+            features. Accepted values are `madx`, `mad-x`, `madng` and `mad-ng`, case-insensitive. If
+            set to `None`, validation will be skipped. See the `tfs.frame.validate` function
```
Good catch. I've modified the `if` statement below to also make sure `validate is not False`, considering that was the way to skip validation in `3.x` and we want users to not have to change that. So if we get `None` or `False`, it is skipped.

For anything else we send to the validation function which is in charge of checking for a correct parameter value.

fsoubelet commented:
And away we go.
